tools/crm_mon.c

```diff
@@ -89,6 +89,7 @@ static pcmk__supported_format_t formats[] = {
 #if CURSES_ENABLED
     CRM_MON_SUPPORTED_FORMAT_CURSES,
 #endif
+    CRM_MON_SUPPORTED_FORMAT_XML,
```

Comment:
One challenge with using a separate format is that you'll have to duplicate (or expose) all of the XML output functions added to lib/pengine and lib/fencing

Reply:
Unless I am missing something, I don't think this will be a problem.  The only differences I have seen so far are the top-level wrapper that finish writes out, lists having a real name vs. "list" as their containing element, and lists having a count at the end vs. not.  What I'm doing in this commit is basically a quasi-OO thing where the crm_mon XML formatter just overrides those couple functions.  Passing that around into the pengine and fencing functions should be okay.  At least, it has previously worked in my testing.

Reply:
How does pcmk__register_messages() know that it should look for "xml" instead of "legacy-xml"?

Reply:
It doesn't know that, but as long as the differences are confined to the primitive functions like finish, begin_list, list_item, etc. then it doesn't matter.  The `pcmk__output_t` instance passed to all those library functions will have its function pointers set to the legacy-xml versions.  Of course, if we discover there are other differences, this will all fall apart and we'll have to do something else.

Reply:
But custom messages aren't function pointers in the struct. They're entries in a hash table, which is created based on the format name. E.g. if someone calls out->message(out, "last-fenced", ...), there won't be any implementation for "legacy-xml".

Reply:
Right, so that will definitely not work.

What I'm saying is that right now, the only differences in the two that I have noticed are at a lower level.  So not last-fenced, but anything in the base formatter structure it uses (like begin_list, or list_item, or whatever) would have an implementation for legacy-xml.

Maybe I am assuming too much about things not being different, though.

Reply:
Sure, the struct methods are straightforward. But the custom messages are where the bulk of the code is. Looking at the definition of pcmk_register_messages(), we might have an easy workaround: it looks like it's possible already to call pcmk__register_messages() twice, first with "xml", then again with "legacy-xml" which will only override messages that have a "legacy-xml" implementation.

Reply:
Hmm, but the format name is taken from out, so that would require replacing "legacy-xml" with "xml", calling pcmk__register_messages() (and its library wrappers), then restoring "legacy-xml" and calling pcmk__register_messages() again to get any local overrides.

Reply:
Do other command line tools support an XML output format?  This could be extremely tedious, but we could add the legacy xml format as a command line option and then everywhere there's a difference, conditionalize it.  This could be tweaked to support multiple legacy XML formats.

Reply:
No, crm_mon is the only one, so an option would be easy.

(Not counting commands that pass along external XML, but those will be handled like stonith_admin --metadata.)

Reply:
Okay, I'll work on that.

Reply:
If you make it a general output option, it would be a good idea to make it hidden so it doesn't clutter help, and it would only be used in crm_mon via pcmk__force_args()

Reply:
Yep, that is the plan.